### PR TITLE
[WIP] draid rebase

### PR DESCRIPTION
**Pull request by ofaaland**
### Motivation and Context
This is a WIP rebase of draid from (most recent at top):
@mmaybee zfs/tree/crayzfs-0.8.x
which was a rebase of:
@don-brady PR #8016
which was a rebase of:
@thegreatgazoo PR #7078

To allow continued community development using a branch based on most recent master.  The rebase is the only improvement.

Very little testing has been done and there are likely new issues resulting from the rebases/merges in addition to whatever issues existed in the earlier PRs.

### Description
Rebasing Notes:
 - Updated to use new types and functions relevant to range trees
 - Background and to-do list as of May 2019 DRAID workshop: https://docs.google.com/document/d/1SDuFWjiAZbsqJSYOt0iO2Dh4RvU-CzcixtQqLW6TuIw/edit#heading=h.zd12b6w407aq

### How Has This Been Tested?
Minor manual testing of draid configs

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by ofaaland** [comment]
@richardelling this is a reminder to push the additional tests you mentioned, when you get a chance, and I'll cherry-pick them here so we get improvements all in one place.  Thanks!

**Comment by mmaybee** [comment]
Very cool Olaf. Please be aware that zfs/tree/crayzfs-0.8.x (which is a rebase from Don's branch) is still a little rough. There are, recently discovered, merge issues in metaslab.c and spa.c.

**Comment by ofaaland** [comment]
I fixed style issues and bad arguments I'd provided to rs_get_start() an an earlier version.

**Comment by mmaybee** [comment]
The merge issues in spa.c are in spa_vdev_attach() (we are not starting a rebuild when we should be). Ther merge issues in metaslab.c are more pervasive: metaslab_block_picker() and metaslab_df_alloc() both need work (because of changes to the algorithms). I can provide details/fixes.

**Comment by ofaaland** [comment]
> The merge issues in spa.c are in spa_vdev_attach() (we are not starting a rebuild when we should be). Ther merge issues in metaslab.c are more pervasive: metaslab_block_picker() and metaslab_df_alloc() both need work (because of changes to the algorithms). I can provide details/fixes.

OK great, thanks!

**Comment by Ornias1993** [comment]
@ofaaland  @mmaybee 
Is there somewhat of a list of what actually needs work on this PR?
- Any specific bugs?
- Any specific issues?
- Just testing?
- More code review?

Being a WIP is nice and all, but I think it makes it hard for other people to start sharing in the work and thought process without a little more background information on the current state of this WIP (after all those years)...

**Comment by ofaaland** [comment]
@Ornias1993 I believe @behlendorf has such a list and is looking for it, for those reasons.  Stay tuned...

**Comment by sjorge** [comment]
Someone should probably also update the feature matrix and link to this PR instead of the abandoned one.

**Comment by ofaaland** [comment]
> Someone should probably also update the feature matrix and link to this PR instead of the abandoned one.

@sjorge I'm happy to see if I can update it.  Where is it?

**Comment by ahrens** [comment]
@ofaaland I already updated it, but it's here: https://docs.google.com/spreadsheets/d/1CFapSYxA5QRFYy5k6ge3FutU7zbAWbaeGN2nKVXgxCI/edit#gid=0

**Comment by Ornias1993** [comment]
@ofaaland Guess i've found the answer to my own question already:
I think you meant @behlendorf was looking for this document?

https://docs.google.com/document/d/1SDuFWjiAZbsqJSYOt0iO2Dh4RvU-CzcixtQqLW6TuIw/edit#heading=h.zd12b6w407aq

Maybe add it to the PR under "Motivation and Context" or "Description", because it answers a LOT of questions.

**Comment by jkiebzak** [comment]
hopefully this is the correct place for this.

I am getting PANICs while trying to create a draid zpool:
```
Message from syslogd@bb-129 at Dec  5 15:14:21 ...
 kernel:[  120.156895] VERIFY3(start == vdev_draid_get_astart(vd, start)) failed (17179869184 == 17179877376)

Message from syslogd@bb-129 at Dec  5 15:14:21 ...
 kernel:[  120.156915] VERIFY3(start == vdev_draid_get_astart(vd, start)) failed (34359738368 == 34359742464)

Message from syslogd@bb-129 at Dec  5 15:14:21 ...
 kernel:[  120.156917] PANIC at vdev_draid.c:478:vdev_draid_check_block()

Message from syslogd@bb-129 at Dec  5 15:14:21 ...
 kernel:[  120.157156] PANIC at vdev_draid.c:478:vdev_draid_check_block()
```

I've tried various draid configurations, all of which PANIC. Here is the most basic configuration
```
draidcfg -p 1 -d 2 -s 1 -n 4 4.nvl
zpool create -f data draid1 cfg=4.nvl /dev/sd{a..d}
```

Here's my system:
```
cat /etc/debian_version
9.11

uname -a
Linux bb-129 4.9.0-11-amd64 #1 SMP Debian 4.9.189-3+deb9u2 (2019-11-11) x86_64 GNU/Linux
```

Here were my installation steps:
```
apt update && apt upgrade

apt install git
git clone https://github.com/zfsonlinux/zfs.git
cd zfs
git pull origin pull/9558/head

apt -y install build-essential autoconf automake libtool gawk alien fakeroot ksh
apt -y install zlib1g-dev uuid-dev libattr1-dev libblkid-dev libselinux-dev libudev-dev
apt -y install libacl1-dev libaio-dev libdevmapper-dev libssl-dev libelf-dev
apt -y install linux-headers-$(uname -r)
apt -y install python3 python3-dev python3-setuptools python3-cffi

sh autogen.sh
./configure --enable-debug
make -s -j$(nproc)
make deb
dpkg -i *.deb

nano /etc/modprobe.d/zfs.conf
  options zfs_vdev_scrub_max_active=10 zfs_vdev_async_write_min_active=4 draid_debug_lvl=5
echo zfs >> /etc/modules
/sbin/modprobe zfs
shutdown -r now
```

**Comment by mmaybee** [comment]
These panics are probably due to the merge issues in the metaslab alloc code path that I mentioned above. I have been meaning to get the fixes pushed, but have not yet found the time. Sorry about the delay.

**Comment by Vlad1mir-D** [comment]
Is it possible to remove or add VDEV into the ZFS pool which contains draid VDEV?

**Comment by richardelling** [comment]
> Is it possible to remove or add VDEV into the ZFS pool which contains draid VDEV?

For top-level vdevs:
* add: yes
* remove: no, due to the same restrictions facing raidz

**Comment by ofaaland** [comment]
Just pushed 2f51d38 Brian ginned up
* rebased on master
* includes fixes by @mmaybee  & co at Cray
* includes fixes by @behlendorf 
* squashed

This is the current best available draid patch that I am aware of

**Comment by Ornias1993** [comment]
This is some significant progress guys, thank you! :)

**Comment by Ornias1993** [comment]
@behlendorf 
I don't think this is going to be in OpenZFS2.0, it might be prudent to just create a placeholder project "OpenZFS2.0" and drop this there.

**Comment by behlendorf** [comment]
We should have an updated version of this ready next week.  But yes, there's still a fair bit of work needed and we don't plan to block the OpenZFS release on this feature.  I've gone ahead and removed it from the project.

**Comment by Ornias1993** [comment]
@behlendorf  I'm looking forward to see the progress made! :)
How big is the progres you guys made as of yet?

File: module/zfs/arc.c [hunk 1]
```diff
@@ -5702,6 +5702,7 @@ arc_read(zio_t *pio, spa_t *spa, const blkptr_t *bp,
 		 * Gracefully handle a damaged logical block size as a
 		 * checksum error.
 		 */
+		ASSERT3U(lsize, <=, spa_maxblocksize);
```
Comment (domag02) [inline module/zfs/arc.c]: Missing function parameter list.
Should be:
`ASSERT3U(lsize, <=, spa_maxblocksize(spa));`

**Comment by behlendorf** [comment]
Decent progress has been made.  I've opened PR #10102 which replaces this PR.

**Pull request closed without merging**